### PR TITLE
role.yaml rights assessment

### LINKED ISSUE #440: role.yaml rights assessment
Operator currently runs with more rights than necessary.
A fine tuning would be good to enhance security.

### PR DESCRIPTION
**Pull request by dmvolod**
Fixes #440

### CODE REVIEW

**Comment by rigazilla** [comment]
We should have a way to run testsuite as an `infinispan-operator` account, to test this kind of PR. I'm looking into

**Comment by dmvolod** [comment]
> We should have a way to run testsuite as an `infinispan-operator` account, to test this kind of PR. I'm looking into

Yes, it's possible but need some manual work. make push-okd4 deploys operator with account support and tests can be run with RUN_LOCAL=false
I did the same steps during tunung this rules

**Comment by rigazilla** [comment]
I have errors with this test procedure:
```
oc new-project namespace-for-testing
make test RUN_LOCAL_OPERATOR=false
history 
```

**Comment by dmvolod** [comment]
> history

Looking to it. Something wrong with registry detection.

**Comment by dmvolod** [comment]
> I have errors with this test procedure:
> 
> ```
> make push-okd4
> oc new-project namespace-for-testing
> make test RUN_LOCAL_OPERATOR=false
> history 
> ```

Oh, looks like we need more documented on this procedure.
The namespace need to be created before, as push-okd4 creates an ImageStream there. Also following env variables need to be set 
export RUN_LOCAL_OPERATOR=false
export TESTING_NAMESPACE=
export PROJECT_NAME=
export OC_USER=
export EXPOSE_SERVICE_TYPE=LoadBalancer

**Comment by rigazilla** [comment]
I've been able to run `make test` as an `infinispan-operator` account. 
Code needs some small changes for this, I'm wondering if we should put some effort in this and run the test as `infinispan-operator` also in CI.

**Comment by rigazilla** [comment]
Also I'm surprised that `persistentvolumeclaims` access is not necessary... probably because is PV binding is driven by the pod machinery (?)

**Comment by dmvolod** [comment]
> Also I'm surprised that `persistentvolumeclaims` access is not necessary... probably because is PV binding is driven by the pod machinery (?) 

We don't need it right now, but will need in the current PR's. I'm planning to add these rights

**Comment by dmvolod** [comment]
@rigazilla what do you think to add some synthetic tests for cover all required Roles and ClusterRoles by ServiceAccount?

**Comment by rigazilla** [comment]
> @rigazilla what do you think about to add some synthetic tests for cover all required Roles and ClusterRoles by ServiceAccount?

sounds good, I think they could save us from regression

**Review by ryanemerson — APPROVED** [review]

**Comment by dmvolod** [comment]
> sounds good, I think they could save us from regression

I add call of the infinispan.GetNodePortServiceHostPort here, StorageClass lookup will be in the corresponding PR
Let me know, if you can see any other rights not covered by integration tests.
Hope, @ryanemerson will cover Pod's CRUD rights with the backup/restore tests.

File: pkg/controller/infinispan/xsite.go [hunk 1]
```diff
@@ -178,7 +178,7 @@ func getNodePortServiceHostPort(service *corev1.Service, k *kube.Kubernetes, log
 
 	if err != nil || len(workerList.Items) == 0 {
 		// Fallback selecting everything
-		err = k.Client.List(context.TODO(), workerList, nil)
+		err = k.Client.List(context.TODO(), workerList, &client.ListOptions{})
```
Comment (dmvolod) [inline pkg/controller/infinispan/xsite.go:181]: @rigazilla please pay attention, that xsite regression have found and need to be backported to 2.0.x, as I'm not sure that we need to backport whole roles assessment to the 2.0.x

Comment (rigazilla) [reply pkg/controller/infinispan/xsite.go:181]: @dmvolod good spot! It's probably better to have a separate issue/bug/pr for this and a 2.0.x backport

Comment (dmvolod) [reply pkg/controller/infinispan/xsite.go:181]: > good spot! It's probably better to have a separate issue/bug/pr for this and a 2.0.x backport

Yes, fixed in #588

File: test/e2e/main_test.go [hunk 2]
```diff
@@ -112,6 +115,12 @@ func TestNodeStartup(t *testing.T) {
 	waitForPodsOrFail(spec, 1)
 }
 
+// Run some functions for testing rights not covered by integration tests
```
Comment (rigazilla) [inline test/e2e/main_test.go:118]: Tests don't run in the `infinispan-operator` account, only the operator does.
We need another kubernetes object to verify roles directly from test.
I'm adding a commit with a fix

`testKube.Kubernetes` runs in the caller context (`kind-kind` in CI), to test the `infinispan-operator` role from we need another kubernetes object.
Let me add a commit

**Comment by dmvolod** [comment]
Sorry, @rigazilla , I occasionally delete your commit with force psuh

**Comment by rigazilla** [comment]
> Sorry, @rigazilla , I accidentally delete your commit with force push

no worries, committing again

File: test/e2e/main_test.go (hunk 2 above)
Comment (dmvolod) [reply test/e2e/main_test.go:119]: Thanks @rigazilla . Looks very good.

**Review by rigazilla — APPROVED** [review]

**Comment by rigazilla** [comment]
merged, thanks @dmvolod

**Pull request merged**